### PR TITLE
Hitec-Airspeed: add support for CAN/I2C selection switch

### PR DESCRIPTION
**Pull request by nathanhitec**
PR for adding a selection switch and a corresponding LED for the Hitec-Airspeed board. This board contains a CAN and I2C interface for the airspeed sensor. This switch will disallow the possibility of  having two I2C masters on the same bus when a device is connected to it's I2C interface.  When this switch is closed, the CAN_I2C_SEL pin is pulled low and the airspeed sensor is initialized. When the switch is open, the CAN_I2C_SEL pin is pulled high and the airspeed sensor is not initialized. This allows another master device to initialize and use the airspeed sensor without interference from the STM32 on the airspeed board.

### CODE REVIEW

**Comment by nathanhitec** [comment]
It doesn't, what would be the appropriate way to do that? Right now I'm looking at using palSetLineMode(HAL_GPIO_PIN_I2C1_SCL, PAL_MODE_INPUT)
palSetLineMode(HAL_GPIO_PIN_I2C1_SDA, PAL_MODE_INPUT)
PAL_MODE_INPUT is said to be a regular high-Z pad mode i.e. floating. Would this be sufficient?  I just checked and it correctly changes the MODER bits for each line from alt function mode to input mode. Also, if this is sufficient, is it necessary to wrap these in an #ifdef ? There are a couple of nested #ifdefs already and it's essentially implicit that these i2c lines exist so that's why I'm asking. Thanks.

**Comment by tridge** [comment]
> palSetLineMode(HAL_GPIO_PIN_I2C1_SCL, PAL_MODE_INPUT)
> palSetLineMode(HAL_GPIO_PIN_I2C1_SDA, PAL_MODE_INPUT)

that looks good, but I'd do a few extra things. We don't want it to be airspeed specific, instead put it in ChibiOS::I2CBus::clear_all() in AP_HAL_ChibiOS/I2CDevice.cpp and check for HAL_GPIO_PIN_GPIO_CAN_I2C1_SEL, HAL_GPIO_PIN_GPIO_CAN_I2C2_SEL etc for up to 4 buses. That way it can be used for both AP_Periph and flight firmware.

**Comment by hendjoshsr71** [comment]
@nathanhitec  I have fixed up the merge conflict as it was an easy one and split up your commits per library per our guidelines.

See this documentation here about how to split up and squash commits.  https://ardupilot.org/dev/docs/git-interactive-rebase.html

**Comment by nathanhitec** [comment]
@hendjoshsr71 Ok, thank you I appreciate it. Is there anything more I should do at this point?

**Review by peterbarker — CHANGES_REQUESTED** [review]
`Tools` commit is actually in AP_HAL_ChibiOS` and should be renamed.  Or squashed into the other commit?

**Comment by nathanhitec** [comment]
Think I got it right thanks again for taking a look at this guys.

File: libraries/AP_HAL_ChibiOS/I2CDevice.cpp [hunk 1]
```diff
@@ -105,6 +105,45 @@ void I2CBus::clear_all()
...
+        palSetLineMode(sda_line, PAL_MODE_INPUT);
+        palSetLineMode(scl_line, PAL_MODE_INPUT);
+    }
+}
+
+
+/*
+  Check enabled I2C/CAN select pins against check_pins bitmask
+ */
+uint8_t I2CBus::check_select_pins(uint8_t check_pins)
```
Comment (hendjoshsr71) [inline libraries/AP_HAL_ChibiOS/I2CDevice.cpp:126]: This function outputs an `uint8_t`, but you have a boolean output at the return being converted

```suggestion
bool I2CBus::check_select_pins(uint8_t check_pins)
```

File: libraries/AP_HAL_ChibiOS/I2CDevice.h [hunk 2]
```diff
@@ -52,6 +52,8 @@ class I2CBus : public DeviceBus {
     static void clear_all(void);
     static void clear_bus(uint8_t busidx);
     static uint8_t read_sda(uint8_t busidx);
+    static uint8_t check_select_pins(uint8_t check_pins);
```
Comment (hendjoshsr71) [inline libraries/AP_HAL_ChibiOS/I2CDevice.h:55]:
```suggestion
    static bool check_select_pins(uint8_t check_pins);
```

File: Tools/AP_Periph/AP_Periph.cpp [hunk 3]
```diff
@@ -180,10 +181,24 @@ void AP_Periph_FW::init()
 #endif
 
 #ifdef HAL_PERIPH_ENABLE_AIRSPEED
-    if (airspeed.enabled()) {
-        // Note: logging of ARSPD is not enabled currently. To enable, call airspeed.set_log_bit();
+    if (airspeed.enabled()){
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+        uint8_t pins_enabled = ChibiOS::I2CBus::check_select_pins(0x01);
```
Comment (hendjoshsr71) [inline Tools/AP_Periph/AP_Periph.cpp:186]:
```suggestion
        const bool pins_enabled = ChibiOS::I2CBus::check_select_pins(0x01);
```

**Review by hendjoshsr71 — CHANGES_REQUESTED** [review]
I personally, think your hwdef change should be its own commit. It is a separate addition from the functionality addition. But not a big deal either way.

File: Tools/AP_Periph/AP_Periph.cpp (hunk 3 above)
Comment (nathanhitec) [reply Tools/AP_Periph/AP_Periph.cpp]: Edited changes in by rebasing

File: libraries/AP_HAL_ChibiOS/I2CDevice.cpp (hunk 1 above)
Comment (nathanhitec) [reply libraries/AP_HAL_ChibiOS/I2CDevice.cpp]: edited changes in by rebasing

File: libraries/AP_HAL_ChibiOS/I2CDevice.h (hunk 2 above)
Comment (nathanhitec) [reply libraries/AP_HAL_ChibiOS/I2CDevice.h]: edited changes in by rebasing

**Comment by hendjoshsr71** [comment]
@nathanhitec  This looks good to me from a code style perspective, but it is not an area of code I know well enough to put an approval on it. 

I labeled this for DevCallEU to get looked at by Tridge and others.

**Review by tridge — APPROVED** [review]

**Pull request merged**